kinesis.go

```diff
@@ -14,86 +13,90 @@ func init() {
 	router.AdapterFactories.Register(NewKinesisAdapter, "kinesis")
 }
 
+var (
+	ErrMissingTagKey   = errors.New("the tag key is empty, check your template KINESIS_STREAM_TAG_KEY.")
+	ErrMissingTagValue = errors.New("the tag value is empty, check your template KINESIS_STREAM_TAG_VALUE.")
+)
+
 type KinesisAdapter struct {
-	Client     *kinesis.Kinesis
-	Drainers   map[string]*Drainer
-	Launched   map[string]bool
+	Streams    map[string]*Stream
 	StreamTmpl *template.Template
-	mutex      sync.Mutex
+	TagTmpl    *template.Template
+	PKeyTmpl   *template.Template
 }
 
 func NewKinesisAdapter(route *router.Route) (router.LogAdapter, error) {
-	drainers := make(map[string]*Drainer)
-	launched := make(map[string]bool)
-	client := kinesis.New(&aws.Config{})
-	tmpl, err := compileTmpl("KINESIS_STREAM_TEMPLATE")
+	sTmpl, err := compileTmpl("KINESIS_STREAM_TEMPLATE")
 	if err != nil {
 		return nil, err
 	}
 
+	tagTmpl, err := compileTmpl("KINESIS_STREAM_TAG_VALUE")
+	if err != nil {
+		return nil, err
+	}
+
+	pTmpl, err := compileTmpl("KINESIS_PARTITION_KEY_TEMPLATE")
+	if err != nil {
+		return nil, err
+	}
+
+	streams := make(map[string]*Stream)
+
 	return &KinesisAdapter{
-		Client:     client,
-		Drainers:   drainers,
-		Launched:   launched,
-		StreamTmpl: tmpl,
+		Streams:    streams,
+		StreamTmpl: sTmpl,
+		TagTmpl:    tagTmpl,
+		PKeyTmpl:   pTmpl,
 	}, nil
 }
 
 func (a *KinesisAdapter) Stream(logstream chan *router.Message) {
-	for {
-		m, open := <-logstream
-		if !open {
-			log.Println("kinesis: Channel is closed, flushing all the buffers")
-			logErrs(a.FlushAll())
-			break
-		}
-
-		streamName, err := executeTmpl(a.StreamTmpl, m)
+	for m := range logstream {
+		sn, err := executeTmpl(a.StreamTmpl, m)
 		if err != nil {
 			logErr(err)
```

Comment:
We should make logErr a public variable inside this package. Then we can set it from [honeybadger.go](https://github.com/remind101/logspout/blob/master/honeybadger.go) to report the error to honeybadger:

```go
var ErrorHandler = func(err error) {
        // Log error by default
}
```

```go
kinesis.ErrorHandler = func(err error) {
        honeybadger.Notify(err)
}
```